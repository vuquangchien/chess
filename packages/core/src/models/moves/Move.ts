import {Cell} from '../Cell';
import {PieceColor} from '../PieceColor';
import {PieceColors} from '../Piece';

export class Move {
  public from: Cell;
  public to: Cell;
  public player: PieceColor | undefined;

  constructor(from: Cell, to: Cell) {
    this.from = from;
    this.to = to;
    this.player = from.piece?.getColor();
  }
  public fileDistance(): number {
    return Math.abs(this.from.colIndex - this.to.colIndex);
  }
  public rankDistance(): number {
    return Math.abs(this.from.rowIndex - this.to.rowIndex);
  }

  public isDiagonal(): boolean {
    return this.fileDistance() === this.rankDistance();
  }
  public isVertical(): boolean {
    return this.from.colIndex === this.to.colIndex;
  }
  public isHorizontal(): boolean {
    return this.from.rowIndex === this.to.rowIndex;
  }
  public isKnightMove(): boolean {
    return (
      (this.fileDistance() === 2 && this.rankDistance() === 1) ||
      (this.fileDistance() === 1 && this.rankDistance() === 2)
    );
  }
  public isForward(): boolean {
    return (
      (this.player === PieceColors.WHITE &&
        this.from.rowIndex > this.to.rowIndex) ||
      (this.player === PieceColors.BLACK &&
        this.from.rowIndex < this.to.rowIndex)
    );
  }
  public isStraight(): boolean {
    return this.isVertical() || this.isHorizontal();
  }
  public isSimpleMove(): boolean {
    return this.to.piece === null;
  }
  public isCaptureMove(): boolean {
    return this.to.piece !== null;
  }
  cellsBetween(): string[] {
    if (this.isKnightMove()) {
      return [];
    }
    return [];
  }
  toString(): string {
    return `${this.from.piece?.color}:${this.from.name} -> ${this.to.name}`;
  }
}
