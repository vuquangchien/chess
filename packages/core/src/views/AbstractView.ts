import {IView} from '../framework/IView';
import {IEngine} from './IEngine';

export abstract class AbstractView<T> implements IView<T> {
  engine!: IEngine;
  children: AbstractView<unknown>[] = [];
  controller: T;

  public constructor(controller: T) {
    this.children = [];
    this.controller = controller;
    this.init();
  }
  init(): void {}

  setEngine(engine: IEngine): void {
    this.engine = engine;
    this.children.forEach(child => child.setEngine(engine));
  }
  addChild(child: AbstractView<unknown>): void {
    this.children.push(child);
  }

  addChildren(children: AbstractView<unknown>[]): void {
    this.children.push(...children);
  }

  renderChildren(): string {
    let result = '';
    for (const child of this.children) {
      result += child.render();
    }
    return result;
  }

  abstract render(): string;
}
