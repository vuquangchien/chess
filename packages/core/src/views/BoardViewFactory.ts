import {Board} from '../models/Board';
import {BoardViewController} from './BoardViewController';
import {BoardView} from './BoardView';

export class BoardViewFactory {
  public static fromBoard(board: Board): BoardView {
    return new BoardView(new BoardViewController(board));
  }
}
