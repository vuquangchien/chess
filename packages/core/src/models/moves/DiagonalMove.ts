import {NumberIterator} from '../helpers/NumberIterator';
import {Cell} from '../Cell';
import {Move} from './Move';

export class DiagonalMove extends Move {
  cellsBetween(): string[] {
    const files = [];
    const it = NumberIterator.createIterator(
      this.from.colIndex,
      this.to.colIndex,
      {
        excludeFrom: true,
        excludeTo: true,
      }
    );
    while (it.hasNext()) {
      files.push(it.next());
    }
    const ranks: number[] = [];
    const it2 = NumberIterator.createIterator(
      this.from.rowIndex,
      this.to.rowIndex,
      {
        excludeFrom: true,
        excludeTo: true,
      }
    );
    while (it2.hasNext()) {
      ranks.push(it2.next());
    }
    return files.map((file, index) => {
      return new Cell(file, ranks[index], null).name;
    });
  }
}
