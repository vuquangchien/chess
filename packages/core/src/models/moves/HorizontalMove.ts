import {NumberIterator} from '../helpers/NumberIterator';
import {Cell} from '../Cell';
import {Move} from './Move';

export class HorizontalMove extends Move {
  cellsBetween(): string[] {
    const rank = this.from.rowIndex;
    const files = [];
    const it = NumberIterator.createIterator(
      this.from.colIndex,
      this.to.colIndex,
      {
        excludeFrom: true,
        excludeTo: true,
      }
    );
    while (it.hasNext()) {
      files.push(it.next());
    }
    return files.map(file => new Cell(file, rank, null).name);
  }
}
