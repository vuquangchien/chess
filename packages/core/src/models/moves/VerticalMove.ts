import {NumberIterator} from '../helpers/NumberIterator';
import {Cell} from '../Cell';
import {Move} from './Move';

export class VerticalMove extends Move {
  cellsBetween(): string[] {
    const file = this.from.colIndex;
    const ranks = [];
    const it = NumberIterator.createIterator(
      this.from.rowIndex,
      this.to.rowIndex,
      {
        excludeFrom: true,
        excludeTo: true,
      }
    );
    while (it.hasNext()) {
      ranks.push(it.next());
    }
    return ranks.map(rank => new Cell(file, rank, null).name);
  }
}
