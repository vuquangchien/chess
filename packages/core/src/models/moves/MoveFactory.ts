import {Cell} from '../Cell';
import {VerticalMove} from './VerticalMove';
import {HorizontalMove} from './HorizontalMove';
import {DiagonalMove} from './DiagonalMove';
import {Move} from './Move';
import {Board} from '../Board';
import {PieceColor} from '../PieceColor';

export class MoveFactory {
  static fromCellNames(board: Board, from: string, to: string): Move {
    const fromCell = board.getByCellName(from);
    const toCell = board.getByCellName(to);
    return MoveFactory.createMove(fromCell, toCell);
  }
  static createMove(from: Cell, to: Cell): Move {
    if (from.colIndex === to.colIndex) {
      return new VerticalMove(from, to);
    }
    if (from.rowIndex === to.rowIndex) {
      return new HorizontalMove(from, to);
    }
    if (
      Math.abs(from.colIndex - to.colIndex) ===
      Math.abs(from.rowIndex - to.rowIndex)
    ) {
      return new DiagonalMove(from, to);
    }
    return new Move(from, to);
  }
  static fromPgn(board: Board, pgn: string, color: PieceColor): Move {
    const to = pgn;
    const toCell = board.getByCellName(to);
    const allPossibleMoves = board.getAllPossibleMoves();
    const possibleMoves = allPossibleMoves.filter(move => {
      return move.to.name === toCell.name && move.player === color;
    });
    if (possibleMoves.length === 0) {
      throw new Error(`No possible moves for ${to}, player ${color}`);
    }
    if (possibleMoves.length === 1) {
      return possibleMoves[0];
    }
    throw new Error(`Ambiguous move ${to}`);
  }
}
