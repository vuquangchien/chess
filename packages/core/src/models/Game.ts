import {GameMove} from './GameMove';
import {BoardFactory} from './BoardFactory';
import {Board} from './Board';
import {Cell} from './Cell';
import {MoveFactory} from './moves/MoveFactory';

export class Game {
  public initialBoard: Board;
  public moves: GameMove[] = [];

  constructor(
    board: Board = BoardFactory.createBoardWithPieces(),
    moves: GameMove[] = []
  ) {
    this.initialBoard = board;
    this.moves = moves;
  }

  get lastMove(): GameMove | null {
    return this.moves.length > 0 ? this.moves[this.moves.length - 1] : null;
  }

  public get board(): Board {
    return this.lastMove?.endBoard ?? this.initialBoard;
  }
  public movePieceByName(from: string, to: string): void {
    const fromCell = this.board.getByCellName(from);
    const toCell = this.board.getByCellName(to);
    this.movePiece(fromCell, toCell);
  }
  public movePiece(from: Cell, end: Cell): void {
    const move = MoveFactory.createMove(from, end);
    const currentBoard = this.lastMove?.endBoard ?? this.initialBoard;
    const startBoard = currentBoard.clone();
    currentBoard.movePiece(move.from, move.to);
    const endBoard = currentBoard.clone();
    this.moves.push(new GameMove(startBoard, endBoard, move));
  }
}
