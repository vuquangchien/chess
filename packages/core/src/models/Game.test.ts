import {Game} from './Game';
import {BoardViewFactory} from '../views/BoardViewFactory';
import {MoveFactory} from './moves/MoveFactory';

describe('Game', () => {
  it('should init game', () => {
    const game = new Game();
    expect(game.initialBoard).toBeDefined();
    expect(game.moves).toEqual([]);
    expect(
      BoardViewFactory.fromBoard(game.initialBoard).render()
    ).toMatchSnapshot();
  });
  it('should move piece', () => {
    const game = new Game();
    game.movePieceByName('a2', 'a3');
    expect(game.moves.length).toEqual(1);
    expect(
      BoardViewFactory.fromBoard(game.initialBoard).render()
    ).toMatchSnapshot();
  });
});
