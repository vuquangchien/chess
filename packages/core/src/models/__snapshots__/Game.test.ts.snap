// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should init game 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □♜□♞□♝□♛□♚□♝□♞□♜
□7 □♟□♟□♟□♟□♟□♟□♟□♟
□6 □ □ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □ □ □ □ □ 
□3 □ □ □ □ □ □ □ □ 
□2 □♙□♙□♙□♙□♙□♙□♙□♙
□1 □♖□♘□♗□♕□♔□♗□♘□♖
"
`;

exports[`Game should move piece 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □♜□♞□♝□♛□♚□♝□♞□♜
□7 □♟□♟□♟□♟□♟□♟□♟□♟
□6 □ □ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □ □ □ □ □ 
□3 □♙□ □ □ □ □ □ □ 
□2 □ □♙□♙□♙□♙□♙□♙□♙
□1 □♖□♘□♗□♕□♔□♗□♘□♖
"
`;
